myia/cconv.py

```diff
@@ -0,0 +1,165 @@
+from .anf_ir import ANFNode, Constant, Graph
+from .anf_ir_utils import dfs
+
+
+def is_graph(x: ANFNode) -> bool:
+    """Return whether x is a Constant with a Graph value."""
+    return isinstance(x, Constant) and isinstance(x.value, Graph)
+
+
+class ParentProxy:
+    """Represents a graph's immediate parent."""
+
+    def __init__(self, graph: Graph) -> None:
+        """Initialize the ParentProxy."""
+        self.graph = graph
+
+
+class NestingAnalyzer:
+    """
+    Analyzes the nesting structure of a graph.
+
+    Call the `run(root_graph)` method to populate the data structures.
+
+    Attributes:
+        deps: Maps a graph to the set of graphs that it depends on.
+        parents: Maps a graph to its immediate parent or `None` if
+            it has no parent.
+        fvs: Maps a graph to a set of nodes it refers to from other
+            graphs, including nodes nested graphs have to access.
+
+    """
+
+    def __init__(self) -> None:
+        """Initialize the NestingAnalyzer."""
+        self.processed: Set[Graph] = set()
+        self.parents: Dict[Graph, Optional[Graph]] = {}
+        self.fvs: Dict[Graph, Set[ANFNode]] = defaultdict(set)
+
+    def run(self, root: Graph) -> 'NestingAnalyzer':
+        """Run the analysis. Populates `self.parents` and `self.fvs`."""
+        if root in self.processed:
+            return self
+
+        # Initialize temporary structures
+        self.prox: Dict[Graph, ParentProxy] = {}
+        self.graphs: Set[Graph] = set()
+        self.nodes: Set[ANFNode] = set()
+        self.deps: Dict[Graph, Set[Union[Graph, ParentProxy]]] = \
+            defaultdict(set)
+
+        # We get all nodes that could possibly be executed within this
+        # graph.
+        self.nodes = set(dfs(Constant(root), True))
+
+        # We get all graphs possibly accessed.
+        self.graphs = {node.value for node in self.nodes if is_graph(node)}
+        self.processed |= self.graphs
+
+        # We compute the dependencies:
+        # * If a node in graph g has an input from graph g', then g depends
+        #   on g'
+        # * If a node in graph g refers to a graph g', then g depends on
+        #   whatever the parent scope of g' is. That parent scope is
+        #   represented by a ParentProxy.
+        self.compute_deps()
+
+        # We complete the dependency graph by following proxies.
+        self.resolve_proxies()
+
+        # We remove all but the most immediate parent of a graph.
+        self.simplify()
+
+        # We update the parents. ParentProxy instances were removed, so
+        # the typing's correct.
+        new_parents = {g: None if len(gs) == 0 else list(gs)[0]
+                       for g, gs in self.deps.items()}
+        self.parents.update(new_parents)  # type: ignore
+
+        # We find all free variables (could be interleaved with other
+        # phases, probably)
+        self.associate_free_variables()
+
+        return self
+
+    def compute_deps(self) -> None:
```

Comment:
I undertand that your logic has different steps, but you might not want to expose them as public functions.